### PR TITLE
Synonyms API - PUT Synonym Rule request

### PR DESCRIPTION
**Pull request by carlosdelest**
Adds PUT Synoym Rule request:

`PUT _synonyms/<synonym_set_id>/<synonym_rule_id>`

```json
{
   "synonyms": "hello, hi, howdy"
}
```

Result is the same as with PUT synonym set request:

```json
{
    "result": "updated",
    "reload_analyzers_details": {
        "_shards": {
            "total": 2,
            "successful": 1,
            "failed": 0
        },
        "reload_details": [
            {
                "index": "index1",
                "reloaded_analyzers": [
                    "my_analyzer"
                ],
                "reloaded_node_ids": [
                    "vxKVyIRJQuO4DPj8Y70ijg"
                ]
            }
        ]
    }
}
```
`result` can be `updated` or `created` depending on a previous existing synonym rule with the same ID.

This operation refreshes all analyzers, not just the ones that use the synonym set. This is in line with https://github.com/elastic/elasticsearch/pull/96886

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java:122]: Refactored this response into a common `SynonymUpdateResponse`, that will be used from both PUT actions

File: server/src/main/java/org/elasticsearch/action/synonyms/SynonymUpdateResponse.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/action/synonyms/SynonymUpdateResponse.java:23]: (this one)

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:178]: Used a  `DelegatingIndexNotFoundActionListener` class to ensure that when index is not found, it is translated into a `ResourceNotFound`. I'd like to have reused the class from Enterprise Search, but given it's in XPack, I didn't want to push things here.

If anyone can think of a good place for this class to be, I'm up for moving it to a top-level class!

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:243]: I hate the `l1`, `l2` naming. Any other suggestion?

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:277]: I refactored this out as it will be reused by `DELETE` and `GET` actions as well; we need to check for the synonym set existence.

File: rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json
Comment (kderusso) [inline rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json:5]:
```suggestion
      "description": "Creates or updates a synonym rule inside a synonym set"
```

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/50_synonym_rule_put.yml
Comment (kderusso) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/50_synonym_rule_put.yml:44]: Possible additional test case: Create a synonym rule for a synonym set that does not yet exist and validate the expected response.

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java:46]: Nitpick: should this be singular? 

```suggestion
            "synonym",
```

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java:77]: Can't hurt to keep the paranoia going strong 😉 

```suggestion
            if (Strings.isNullOrEmpty(synonymsSetId)) {
```

File: server/src/main/java/org/elasticsearch/action/synonyms/SynonymUpdateResponse.java
Comment (kderusso) [reply server/src/main/java/org/elasticsearch/action/synonyms/SynonymUpdateResponse.java:23]: Nice refactoring 👍

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (kderusso) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:243]: `synonymListener` and `indexListener`?

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:277]:
```suggestion
    private <T> void checkSynonymSetExists(String synonymsSetId, ActionListener<T> listener) {
```

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:292]:
```suggestion
    private void deleteSynonymsSetRules(String resourceName, ActionListener<BulkByScrollResponse> listener) {
```

File: x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java
Comment (kderusso) [inline x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java:105]: I think you still need this, to put a synonyms set?

**Review by kderusso — COMMENTED** [review]
Looking good!

File: rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json:5]: "in" would probably be better, right? (Ah, English prepositions...) 😞

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/50_synonym_rule_put.yml
Comment (carlosdelest) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/50_synonym_rule_put.yml:44]: I think that's included in line 70 😉

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java:46]: This way it has the same name as the field it reads, I think it's ok to leave it as is

File: rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json
Comment (kderusso) [reply rest-api-spec/src/main/resources/rest-api-spec/api/synonym_rule.put.json]: In would be good!

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymRuleAction.java:46]: Here we are dealing with multiple synonyms, right? So "synonyms" is more appropriate.

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:106]: +1 indeed there is no way around it but duplicating _id in the extra field

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:193]: Do we need to use `externalSynonymRuleId` function?  
May be we can add another test to ensure proper ids.

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:178]: I am not sure if it is necessary to introduce this class for this small single change, but I will leave this to you.

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:243]: Why do we need this `checkSynonymSetExists`?   Can we just assume a user want to create new synonymsSet with this new SynonymRule? This adds an extra search request, which I think is not necessary.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:243]: I'd say that could be confusing - in case a user mistypes the synonyms set ID, then instead of an error, she would have a new synonyms set.

I think we should err on the side of caution, and make the user aware that the synonyms set does not exist and needs to be created.

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:193]: > Do we need to use externalSynonymRuleId function?

Now we can remove it, thanks!

> May be we can add another test to ensure proper ids.

What do you have in mind?

File: server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java:122]: You can merge auto-reloading first and I can take care of that later - either reversing the change or making the refactoring after auto-reloading merged

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:234]: Synchronized with @carlosdelest offline, and we decided to keep the current implementation with a search request.  But added a TODO item for GA to deal with empty collections.

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:193]: About the test I was thinking to have a check on synonyms IDs that they don't change (but I see that is already covered by our current yml tests). So no action needed.

**Review by mayya-sharipova — COMMENTED** [review]
@carlosdelest Thanks  as always great work! As it stands now, this PR LGTM. But I would like to have another look after rebasing with the `main` branch. Particularly, interested if we want to add reload analyzers on individual updates.

**Comment by carlosdelest** [comment]
Hi @mayya-sharipova ! I added the reload analyzers response as discussed, and did some refactoring. Please take a look!

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:255]: Refactored the reloading analyzers code into a method

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:367]: I used a generic SynonymsReloadResult record, to avoid having multiple records that just differ on the response type

File: modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/70_synonyms_from_index.yml
Comment (carlosdelest) [inline modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/70_synonyms_from_index.yml:109]: Refactored this test to separate into multiple test cases, and added an individual rule update

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:202]: I've found this double generics are more difficult to follow, not sure if it is worth introducing extra generic just for 2 classes.

File: server/src/main/java/org/elasticsearch/action/synonyms/TransportPutSynonymsAction.java
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/action/synonyms/TransportPutSynonymsAction.java:33]: If we are not using `PutSynonymsAction.Response` class, should we delete it?

**Review by mayya-sharipova — APPROVED** [review]
@carlosdelest Thanks Carlos, as always, great job!  +1 to merge after removing `PutSynonymsAction.Response` class

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:202]: I'm not very satisfied with this either. Probably I'll run through other ideas in separate PRs.

File: server/src/main/java/org/elasticsearch/action/synonyms/TransportPutSynonymsAction.java
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/action/synonyms/TransportPutSynonymsAction.java:33]: Thanks for catching this! Removing

**Pull request merged**
